crates/oxc_linter/src/rules/no_unsafe_negation.rs

```diff
@@ -0,0 +1,142 @@
+
+use crate::{context::LintContext, fixer::Fix, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Unexpected logical not in the left hand side of '{0}' operator")]
+#[diagnostic(severity(warning))]
+struct NoUnsafeNegationDiagnostic(&'static str, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUnsafeNegation {
+    /// true: disallow negation of the left-hand side of ordering relational operators
+    /// false: allow negation of the left-hand side of ordering relational operators (<, >, <=, >=)
+    enforce_for_ordering_relations: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow negating the left operand of relational operators
+    ///
+    /// ### Why is this bad?
+    /// Just as developers might type -a + b when they mean -(a + b) for the negative of a sum,
+    /// they might type !key in object by mistake when they almost certainly mean !(key in object)
+    /// to test that a key is not in an object. !obj instanceof Ctor is similar.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// if (!key in object) {
+    ///   //operator precedence makes it equivalent to (!key) in object
+    ///   //and type conversion makes it equivalent to (key ? "false" : "true") in object
+    /// }
+    /// ```
+    NoUnsafeNegation,
+    correctness
+);
+
+impl Rule for NoUnsafeNegation {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let enforce_for_ordering_relations = value
+            .get(0)
+            .and_then(|config| config.get("enforceForOrderingRelations"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+        Self { enforce_for_ordering_relations }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::BinaryExpression(expr) = node.get().kind() else { return; };
+        if self.should_check(expr.operator) {
+            let Expression::UnaryExpression(left) = &expr.left else { return; };
+            if left.operator == UnaryOperator::LogicalNot {
+                Self::report_with_fix(expr, ctx);
+            }
+        }
+    }
+}
+
+impl NoUnsafeNegation {
+    fn should_check(&self, op: BinaryOperator) -> bool {
+        op.is_relational() || (self.enforce_for_ordering_relations && op.is_compare())
+    }
+
+    /// Precondition:
+    /// expr.left is `UnaryExpression` whose operator is '!'
+    fn report_with_fix(expr: &BinaryExpression, ctx: &LintContext<'_>) {
+        // Diagnostic points at the unexpected negation
+        let diagnostic = NoUnsafeNegationDiagnostic(expr.operator.as_str(), expr.left.span());
+
+        let fix_producer = || {
+            // modify `!a instance of B` to `!(a instanceof B)`
+            let modified_code = {
+                let len_hint = (expr.span.end - expr.span.start + 2) as usize;
+                let mut printer = Printer::new(len_hint, PrinterOptions::default());
```

Comment:
Shall we refactor this out in the next PR?

Reply:
> Shall be refactor this into `LintContext` so we don't have to modify all the code if we change the Printer interface?

Done.

> I wonder if the callback should give us a printer 🤔

Imo Printer is more about the mechanism rather than policy so I'd not put that in the callback parameter. We could wrap a Printer in something like a `FixBuilder` if we need some higher level constructs.